doc/dev/mds_internals/quiesce.rst

```diff
@@ -0,0 +1,90 @@
+protocol will execute the same ``quiesce_subvolume`` operation on each MDS
+rank. This allows each rank which may be authoritative for part of the
+subvolume subtree to lock all inodes it is authoritative for in the subvolume
+``SnapRealm``.
+
+
+Inode Quiescelock
+-----------------
+
+The ``quiescelock`` is a new lock for inodes which supports quiescing I/O.  It
+is a type of superlock where every client or MDS operation which accesses an
+inode lock will also implicitly acquire the ``quiescelock`` (readonly). In
+general, this lock is never held except for reading. When a subtree is
+quiesced, the ``quiesce_subvolume_inode`` internal operation will hold
+``quiescelock`` exclusively, thereby denying the **new** acquisition of any
+other inode lock.  The ``quiescelock`` must be ordered before all other locks
+(see ``src/include/ceph_fs.h`` for ordering) in order to act as this superlock.
+
+The reason for this lock is to prevent an operation from blocking on acquiring
+locks held by ``quiesce_subvolume_inode`` while still holding locks obtained
+during path traversal. Notably, the important locks are the ``snaplock`` and
+``policylock`` obtained via ``Locker::try_rdlock_snap_layout`` on all parents
+of the root inode of the request (the ``ino`` in the ``filepath`` struct). If
+that operation waits with those locks held, then a future ``mksnap`` on the
+subvolume inode will be impossible. (The ``mksnap`` RPC only acquires a wrlock
+(write lock) on the ``snaplock`` for the inode to be snapshotted.)
+
+The way ``quiescelock`` helps prevent this is by being the first **mandatory**
+lock acquired and special handling when it cannot be acquired: all locks held
+by the operation are dropped and the operation waits for the ``quiescelock`` to
+be available.  The lock is mandatory in that all inode locks automatically
+include (add) the ``quiescelock`` when calling ``Locker::acquire_locks``. So
+the expected normal flow is that an operation like ``getattr`` will perform its
+path traversal, acquiring parent and dentry locks, then attempt to acquire
+locks on the inode necessary for the requested client caps. The operation will
+fail to acquire the automatically included ``quiescelock``, add itself to the
+``quiescelock`` wait list, and then drop all held locks.
+
+
+Readable quiesced subvolume
+---------------------------
+
+It may be desirable to allow readers to continue accessing a quiesced
+subvolume. One way to do that is to have a separate superlock (yuck) for read
+access, say ``quiescerlock``. If a "readable" quiesce is performed, then
+``quiescerlock`` is not xlocked by ``quiesce_subvolume_inode``. Read locks on
+other (non-quiesce) locks will acquire a read lock only on ``quiescerlock`` and
+no longer on ``quiescelock``.  Write locks would try to acquire both
+``quiescelock`` and ``quiescerlock`` (since writes may also read).
+
+Ideally, it may be a new lock could be used to handle both cases but no such
```

Comment:
```suggestion
Ideally, it may be a new lock class could be used to handle both cases but no such
```

Reply:
Fixed.

Reply:
We call it a LockType in the code; I used "type" i instead.